File: java/src/main/java/dxjruby/Input.java
package dxjruby;

import java.awt.event.KeyEvent;
import java.awt.event.MouseEvent;
import java.util.List;

import dxjruby.input.MouseEventQueue;
import dxjruby.input.MouseEventQueue.DXJRubyMouseEvent;
import dxjruby.input.InputState;
import dxjruby.input.KeyEventQueue;
import dxjruby.input.KeyEventQueue.DXJRubyKeyEvent;

public class Input {

    private static final InputState state;
    private static final MouseEventQueue mouseEventQueue;
    private static final KeyEventQueue keyEventQueue;

    static {
        state = new InputState();
        mouseEventQueue = new MouseEventQueue();
        keyEventQueue = new KeyEventQueue();
    }

    public static void setMousePosition(final int x, final int y) {
        state.setMousePosition(x, y);
    }

    // --------------------------------
    // keyboard

    public static void updateKeyState(final long t) {
        final List<DXJRubyKeyEvent> evs = keyEventQueue.takeFrameEvents(t);
        state.updateKeyState(evs);
    }

    public static void addToKeyEventQueue(final KeyEvent ev) {
        keyEventQueue.add(ev);
    }

    public static boolean keyPushP(final int dxrubyKeyCode) {
        return state.keyPushP(dxrubyKeyCode);
    }

    public static boolean keyReleaseP(final int dxrubyKeyCode) {
        return state.keyReleaseP(dxrubyKeyCode);
    }

    public static boolean keyDownP(final int dxrubyKeyCode) {
        return state.keyDownP(dxrubyKeyCode);
    }

    // --------------------------------
    // mouse

    public static void updateMouseState(final long t) {
        final List<DXJRubyMouseEvent> evs = mouseEventQueue.takeFrameEvents(t);
        state.updateMouseState(evs);
    }

    public static void addToMouseEventQueue(final MouseEvent ev) {
        mouseEventQueue.add(ev);
    }

    public static boolean mousePushP(final int dxrubyMouseCode) {
        return state.mousePushP(dxrubyMouseCode);
    }

    public static boolean mouseReleaseP(final int dxrubyMouseCode) {
        return state.mouseReleaseP(dxrubyMouseCode);
    }

    public static boolean mouseDownP(final int dxrubyMouseCode) {
        return state.mouseDownP(dxrubyMouseCode);
    }

    @SuppressWarnings("boxing")
    public static int toAwtMouseButton(final int dxrubyMouseCode) {
        switch (dxrubyMouseCode) {
        case DXJRubyMouseCode.M_LBUTTON:
            return MouseEvent.BUTTON1;
        case DXJRubyMouseCode.M_RBUTTON:
            return MouseEvent.BUTTON3;
        case DXJRubyMouseCode.M_MBUTTON:
            return MouseEvent.BUTTON2;
        default:
            throw new RuntimeException(
                    String.format("unknown DXRuby mouse code (%s)", dxrubyMouseCode));
        }
    }

    // --------------------------------

    public static int getMouseX() {
        return state.getMouseX();
    }

    public static int getMouseY() {
        return state.getMouseY();
    }

    // --------------------------------

    private static class DXJRubyMouseCode {
        //                                 dxopal  dxruby
        static final int M_LBUTTON = 1; // 1       0
        static final int M_RBUTTON = 2; // 2       1
        static final int M_MBUTTON = 4; // 4       2
    }

}

File: lib/dxjruby/input.rb
module DXJRuby
  module Input
    module MouseCodes
      M_LBUTTON = 1
      M_RBUTTON = 2
      M_MBUTTON = 4
      # # DXOpal extention
      # M_4TH_BUTTON = 8
      # M_5TH_BUTTON = 16
    end

    ## def self._pressing_keys

    ## # Internal setup for Input class
    ## def self._init(canvas)

    ## # Called on every frame from Window
    ## def self._on_tick

    # Return 1 if 'right', -1 if 'left'
    def self.x(pad_number=0)
      ret = 0
      ret += 1 if key_down?(K_RIGHT)
      ret -= 1 if key_down?(K_LEFT)
      ret
    end

    # Return 1 if 'down', -1 if 'up'
    def self.y(pad_number=0)
      ret = 0
      ret += 1 if key_down?(K_DOWN)
      ret -= 1 if key_down?(K_UP)
      ret
    end

    #
    # Keyboard
    #

    # Return true if the key is being pressed
    def self.key_down?(code)
      j_Input.key_down_p(code.value)
    end

    # Return true if the key is just pressed
    def self.key_push?(code)
      j_Input.key_push_p(code.value)
    end

    # Return true if the key is just released
    def self.key_release?(code)
      j_Input.key_release_p(code.value)
    end

    ## # Set DOM element to receive keydown/keyup event
    ## #
    ## # By default, `window` is set to this (i.e. all key events are
    ## # stolen by DXOpal.) If canvas element is set to this, only key events
    ## # happend on canvas are processed by DXOpal.
    ## def self.keyevent_target=(target)

    ## # Return DOM element set by `keyevent_target=`
    ## def self.keyevent_target

    #
    # Mouse
    #

    ## # (internal) initialize mouse events
    ## def self._init_mouse_events

    # Return position of mouse cursor
    # (0, 0) is the top-left corner of the canvas
    def self.mouse_x
      j_Input.get_mouse_x()
    end
    def self.mouse_y
      j_Input.get_mouse_y()
    end

    class << self
      alias mouse_pos_x mouse_x
      alias mouse_pos_y mouse_y
    end

    # Return true if the mouse button is being pressed
    def self.mouse_down?(mouse_code)
      j_Input.mouse_down_p(mouse_code)
    end

    # Return true if the mouse button is pressed in the last tick
    def self.mouse_push?(mouse_code)
      j_Input.mouse_push_p(mouse_code)
    end

    # Return true if the mouse button is released in the last tick
    def self.mouse_release?(mouse_code)
      j_Input.mouse_release_p(mouse_code)
    end

    #
    # Touch (unsupported)
    #

    private

    def self.j_Input
      Java::dxjruby.Input
    end
  end
end
